### PR TITLE
Make sticky participants work with any diagram top

### PR DESCRIPTION
**Pull request by dyon21**

### CODE REVIEW

File: src/utils/dom.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export function getElementDistanceToTop(element: HTMLElement) {
+  var distanceToTop = 0;
+  var currentElement: HTMLElement | null = element;
+
+  while (currentElement) {
```
Comment (MrCoder) [inline src/utils/dom.ts:5]: Is this a good alternative solution? https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect

File: src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue [hunk 2]
```diff
@@ -52,9 +52,10 @@ export default {
     const translate = computed(() => {
       let top = intersectionTop.value + scrollTop.value
       if (intersectionTop.value > INTERSECTION_OFFSET && store?.state.stickyOffset) top += store?.state.stickyOffset
-      if (top < props.offsetTop + PARTICIPANT_TOP) return 0
-      const diagramHeight = store?.state.diagramElement.clientHeight || 0
-      return Math.min(top - PARTICIPANT_TOP, diagramHeight - PARTICIPANT_HEIGHT) - props.offsetTop
+      const diagramHeight = store?.state.diagramElement?.clientHeight || 0
```
Comment (MrCoder) [inline src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue:55]: consider map `diagramElement` to computed property with vuex `mapState`?

File: src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue [hunk 3]
```diff
@@ -37,8 +37,8 @@ import { computed, ref } from 'vue';
 import useDocumentScroll from '@/functions/useDocumentScroll';
 import useIntersectionTop from '@/functions/useIntersectionTop';
 import { useStore } from 'vuex';
+import { getElementDistanceToTop } from '@/utils/dom';
 
-const PARTICIPANT_TOP = 50  // the distance of participants from the top of the diagram
 const PARTICIPANT_HEIGHT = 70
 const INTERSECTION_OFFSET = 10 // a threshold for judging whether the participant is intersecting with the viewport
```
Comment (MrCoder) [inline src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue:43]: rename to `INTERSECTION_THRESHOLD`?

File: src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue [hunk 4]
```diff
@@ -52,9 +52,10 @@ export default {
     const translate = computed(() => {
       let top = intersectionTop.value + scrollTop.value
       if (intersectionTop.value > INTERSECTION_OFFSET && store?.state.stickyOffset) top += store?.state.stickyOffset
-      if (top < props.offsetTop + PARTICIPANT_TOP) return 0
-      const diagramHeight = store?.state.diagramElement.clientHeight || 0
-      return Math.min(top - PARTICIPANT_TOP, diagramHeight - PARTICIPANT_HEIGHT) - props.offsetTop
+      const diagramHeight = store?.state.diagramElement?.clientHeight || 0
+      const diagramTop = store?.state.diagramElement ? getElementDistanceToTop(store?.state.diagramElement) : 0
+      if (top < props.offsetTop + diagramTop) return 0
+      return Math.min(top - diagramTop, diagramHeight - PARTICIPANT_HEIGHT) - props.offsetTop
```
Comment (MrCoder) [inline src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue:58]: Do we not need to take participant top into consideration at all?
image

File: src/utils/dom.ts (hunk 1 above)
Comment (dyon21) [reply src/utils/dom.ts:5]: I had tried it, but they are not the same thing, the ClientRect will change when the content scroll

File: src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue (hunk 3 above)
Comment (dyon21) [reply src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue:43]: The IntersectionObserver API already have a `threshold` parameter, a variable called `INTERSECTION_THRESHOLD ` may causing confusion, I will renamed it to `INTERSECTION_ERROR_MARGIN`

File: src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue (hunk 4 above)
Comment (dyon21) [reply src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue:58]: It already be considered as the `props.offsetTop`

File: src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue (hunk 2 above)
Comment (dyon21) [reply src/components/DiagramFrame/SeqDiagram/LifeLineLayer/Participant.vue:55]: I think it is not necessary, as we only need this value when we the page is scrolling

**Pull request merged**